feature-toggles/feature-toggles-api/src/main/java/com/duckduckgo/feature/toggles/api/FeatureToggles.kt

```diff
@@ -0,0 +1,144 @@
+            }
+            return@newProxyInstance loadToggleMethod(method)
+        } as T
+    }
+
+    private fun loadToggleMethod(method: Method): Toggle {
+        synchronized(featureToggleCache) {
+            featureToggleCache[method]?.let { return it }
+
+            val defaultValue = try {
+                method.getAnnotation(Toggle.DefaultValue::class.java).defaultValue
+            } catch (t: Throwable) {
+                throw IllegalStateException("Feature toggle methods shall have annotated default value")
+            }
+
+            return ToggleImpl(store, getMethodName(method), defaultValue, appVersionProvider).also { featureToggleCache[method] = it }
+        }
+    }
+
+    private fun validateFeatureInterface(feature: Class<*>) {
+        if (!feature.isInterface) {
+            throw IllegalArgumentException("Feature declarations must be interfaces")
+        }
+        if (!feature.isAnnotationPresent(FeatureName::class.java)) {
+            throw IllegalArgumentException("Feature classes must be annotated with [FeatureName]")
+        }
+    }
+
+    private fun getMethodName(method: Method): String {
+        val featureName = method.declaringClass.getAnnotation(FeatureName::class.java).name
+        if (method.name != "self") return "${featureName}_${method.name}"
+
+        // This can throw but should never happen as it's guarded by codegen as well
+        return featureName
+    }
+
+    private fun validateMethod(method: Method, args: Array<out Any>) {
+        if (method.returnType != Toggle::class.java) {
+            throw IllegalArgumentException("Feature method return types must be Toggle")
+        }
+
+        if (args.isNotEmpty()) {
+            throw IllegalArgumentException("Feature methods must not have arguments")
+        }
+    }
+}
+
+interface Toggle {
+    fun isEnabled(): Boolean
+
+    fun setEnabled(state: State)
+
+    data class State(
+        val enable: Boolean = false,
+        val minSupportedVersion: Int? = null,
+        val enabledOverrideValue: Boolean? = null,
+    )
+
+    interface Store {
+        fun set(key: String, state: State)
+
+        fun get(key: String): State?
+    }
+
+    @Target(AnnotationTarget.FUNCTION)
+    @Retention(AnnotationRetention.RUNTIME)
+    annotation class DefaultValue(
+        val defaultValue: Boolean,
+    )
+}
+
+@Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.RUNTIME)
+annotation class FeatureName(
+    val name: String,
+)
+
+internal class ToggleImpl constructor(
+    private val store: Toggle.Store,
+    private val key: String,
+    private val defaultValue: Boolean,
+    private val appVersionProvider: () -> Int,
+) : Toggle {
+    override fun isEnabled(): Boolean {
+        store.get(key)?.let { state ->
+            return state.enable && appVersionProvider.invoke() >= (state.minSupportedVersion ?: 0)
+        } ?: return defaultValue
+    }
+
+    override fun setEnabled(state: Toggle.State) {
```

Comment:
did you still want to make this more clearly related to overriding? thinking on it since i asked originally, I don't think it's necessary; originally thought it was taking in a boolean but taking in a state should make it obvious enough what the options are

Reply:
I realised that the `setEnabled` is also (rightly) called from codegen, where we won't override anything.
I don't think we have great options here. We'll have to leave it to devs to understand they should never call this method from prod code